### DISCUSSION TITLE
Using Controller with tailwindlabs/headlessui Listbox

### DISCUSSION THREAD

**Discussion by miller-productions** [react-hook-form/react-hook-form#3504] [Q&A] [answered]
Hi there,

I'm trying to use Controller with [tailwindlabs/headlessui Listbox](https://github.com/tailwindlabs/headlessui/tree/develop/packages/%40headlessui-react#listbox-select) but seem to be a bit stumped.

I have followed the react-hook-form examples on using Controller, and in this scenario I am using the `render` approach.

Here is a code snippet that is based around an example provided by the tailwindlabs/headlessui team...

```javascript
import React, { useState } from "react";
import { Listbox, Transition } from "@headlessui/react";
import { useForm, Controller } from "react-hook-form";

const people = [
  "Wade Cooper",
  "Arlene Mccoy",
  "Devon Webb",
  "Tom Cook",
  "Tanya Fox",
  "Hellen Schmidt",
  "Caroline Schultz",
  "Mason Heaney",
  "Claudie Smitham",
  "Emil Schaefer"
];

export default function () {
  const [selectedPerson, setSelectedPerson] = useState(people[0]);
  const { control, handleSubmit } = useForm();

  const onSubmit = (data) => {
    alert(JSON.stringify(data));
  };

  return (
    <form onSubmit={handleSubmit(onSubmit)}>
      <div className="flex items-center justify-center p-12">
        <div className="w-full max-w-xs mx-auto">
        <Controller
          control={control}
          defaultValue=""
          name="selectedPerson"
          render={() => (    // <---------- assumption: I would need to pass in onChange here
          <Listbox
            as="div"
            className="space-y-1"
            value={selectedPerson}
            onChange={setSelectedPerson}  // <----- then probably need to do something with onChange from render props here
          >
            {({ open }) => (
              <>
                <Listbox.Label className="block text-sm leading-5 font-medium text-gray-700">
                  Assigned to
                </Listbox.Label>
                <div className="relative">
                  <span className="inline-block w-full rounded-md shadow-sm">
                    <Listbox.Button className="cursor-default relative w-full rounded-md border border-gray-300 bg-white pl-3 pr-10 py-2 text-left focus:outline-none focus:shadow-outline-blue focus:border-blue-300 transition ease-in-out duration-150 sm:text-sm sm:leading-5">
                      <span className="block truncate">{selectedPerson}</span>
                      <span className="absolute inset-y-0 right-0 flex items-center pr-2 pointer-events-none">
                        <svg
                          className="h-5 w-5 text-gray-400"
                          viewBox="0 0 20 20"
                          fill="none"
                          stroke="currentColor"
                        >
                          <path
                            d="M7 7l3-3 3 3m0 6l-3 3-3-3"
                            strokeWidth="1.5"
                            strokeLinecap="round"
                            strokeLinejoin="round"
                          />
                        </svg>
                      </span>
                    </Listbox.Button>
                  </span>

                  <Transition
                    show={open}
                    leave="transition ease-in duration-100"
                    leaveFrom="opacity-100"
                    leaveTo="opacity-0"
                    className="absolute mt-1 w-full rounded-md bg-white shadow-lg"
                  >
                    <Listbox.Options
                      static
                      className="max-h-60 rounded-md py-1 text-base leading-6 shadow-xs overflow-auto focus:outline-none sm:text-sm sm:leading-5"
                    >
                      {people.map((person) => (
                        <Listbox.Option key={person} value={person}>
                          {({ selected, active }) => (
                            <div
                              className={`${
                                active
                                  ? "text-white bg-blue-600"
                                  : "text-gray-900"
                              } cursor-default select-none relative py-2 pl-8 pr-4`}
                            >
                              <span
                                className={`${
                                  selected ? "font-semibold" : "font-normal"
                                } block truncate`}
                              >
                                {person}
                              </span>
                              {selected && (
                                <span
                                  className={`${
                                    active ? "text-white" : "text-blue-600"
                                  } absolute inset-y-0 left-0 flex items-center pl-1.5`}
                                >
                                  <svg
                                    className="h-5 w-5"
                                    xmlns="http://www.w3.org/2000/svg"
                                    viewBox="0 0 20 20"
                                    fill="currentColor"
                                  >
                                    <path
                                      fillRule="evenodd"
                                      d="M16.707 5.293a1 1 0 010 1.414l-8 8a1 1 0 01-1.414 0l-4-4a1 1 0 011.414-1.414L8 12.586l7.293-7.293a1 1 0 011.414 0z"
                                      clipRule="evenodd"
                                    />
                                  </svg>
                                </span>
                              )}
                            </div>
                          )}
                        </Listbox.Option>
                      ))}
                    </Listbox.Options>
                  </Transition>
                </div>
              </>
            )}
          </Listbox>
          )}
          />
        </div>
      </div>
      <div className="flex justify-center">
      <input type="submit" className="py-2 px-4 border border-transparent text-sm leading-5 font-medium rounded-md text-white bg-indigo-600 hover:bg-indigo-500 focus:outline-none focus:border-indigo-700 focus:ring focus:ring-indigo-300 focus:ring-opacity-70 active:bg-indigo-700 transition duration-150 ease-in-out" />
      </div>
    </form>
  );
}
```

I've tried various combinations of passing in and using onChange via render props but I can't seem to set/get a value for the LIstbox.

Wondering if anyone might have a suggestion?

**Comment by miller-productions** [comment, marked as answer]
I think I figured it out...

```javascript
render={({onChange}) => (
  <Listbox
    as="div"
    className="space-y-1"
    value={selectedPerson}
    onChange={(e) => {
      onChange(e);
      setSelectedPerson(e);
    }}
  >
```

**Reply by faithfulojebiyi** [reply]
onChange  is not a function thats the error it returns

**Reply by jeremiasjutz** [reply]
It should be
```jsx
render={({ field: { onChange } }) =>  (
  <Listbox
    as="div"
    className="space-y-1"
    value={selectedPerson}
    onChange={(e) => {
      onChange(e);
      setSelectedPerson(e);
    }}
  >
```

**Reply by faithfulojebiyi** [reply]
I already fixed it a long time... I think it had to do with the version

**Reply by LeonardoNatale** [reply]
Leveraging `watch` you can get rid of `const [selectedPerson, setSelectedPerson] = useState(people[0]);`

```typescript
const { control, handleSubmit, watch } = useForm({defaultValues: {person: people[0]}});
const selectedPerson = watch(selectedPerson)
```

**Reply by matthewgisonno** [reply]
You shouldn't need any of that... by spreading the `field` object into the Listbox component, it will pass the value, onChange, etc to handle all that.

**Comment by jason-nivelo** [comment]
How would I update this component to have a properly linked reference? I have almost the same exact code, but on form validation the Controlled components do not auto-focus

**Reply by matthewgisonno** [reply]
You either need to forward the ref if you're using a wrapper component and then attach that `ref` to the `Listbox.Button` component.  Or if you're rendering the `Listbox` component directly in the `Controller` rendering prop, you can destructure the `field` object to pull out the `ref` and attach it to the `Listbox.Button`...  Something like this:

```js
<Controller
      control={control}
      defaultValue=""
      name="location"
      rules={{ required: true }}
      render={({ field: { ref, ...rest} }) => (
        // The ...rest will contain name, value, onChange, onBlur
        <Listbox {...rest} as="div" className="space-y-1">
          {({ value }) => (
            <>
              <Listbox.Label className="block text-sm leading-5 font-medium text-gray-700">
                Assigned to
              </Listbox.Label>
              <Listbox.Button ref={ref} className="some-class">
                {value}
              </Listbox.Button>
            </>
          )}
        </Listbox>
      )}
    />
```

**Comment by matthewgisonno** [comment]
Hopefully this isn't too hard to follow... I have multiple components with rendering props, that is why I have an array of onChange callbacks

In my form:
```js
  const {
    register,
    handleSubmit,
    control,
    formState: { errors }
  } = useForm();

const isLocationFieldInvalid = 'location' in errors;
```

return:
```js
              <Controller
                control={control}
                defaultValue=""
                name="location"
                rules={{ required: true }}
                render={({ field }) => (
                  <SelectMenu
                    {...rest}
                    {...field}
                    id="location"
                    className={classnames(rest.className)}
                    options={generateFormattedMetroList()}
                    colorScheme={colorScheme}
                    ariaDescribedBy={
                      isLocationFieldInvalid ? 'locationError' : null
                    }
                  />
                )}
              />

          {isLocationFieldInvalid && (
            <FieldErrorMessage id="locationError">
              {getValidationErrorMessage(errors.location)}
            </FieldErrorMessage>
          )}
```

Here is my `SelectMenu` wrapper component in full... You can see I attached the `ref` from the `forwardRef` to the `Listbox.Button` component, which allows the button to receive focus during form validation:

```js
import { Listbox, Transition } from '@headlessui/react';
import classnames from 'classnames';
import PropTypes from 'prop-types';
import React, { Fragment, forwardRef, useCallback } from 'react';

import {
  CheckmarkIcon,
  ChevronDownIcon,
  ChevronUpIcon,
  Icon,
  iconSizes
} from '@/basics/icons';
import { colorSchemePropType } from '@/lib/propTypes';

import styles from './selectMenu.module.css';

// SelectMenu component is a "controlled" component via react-hook-form.
// It is controlled via the Controller component, which provides a `field` object that gets spread into this component.
// The Controller `field` provides this component with `onChange`, `onBlur`, `value`, `name`, and `ref`.
export const SelectMenu = forwardRef(function SelectMenu(
  {
    id,
    name,
    value,
    options = [],
    isRequired = false,
    isDisabled = false,
    isInvalid = false,
    ariaDescribedBy,
    onBlur,
    onChange,
    colorScheme,
    className,
    isScrollable,
    isFitContent,
    ...rest
  },
  ref
) {
  const buttonClassNames = classnames(styles.selectMenuButton, className, {
    [styles[colorScheme]]: colorScheme,
    [styles.invalid]: isInvalid
  });

  return (
    <Listbox
      as="div"
      by="id"
      name={name}
      value={value}
      onBlur={onBlur}
      onChange={onChange}
      disabled={isDisabled}
      required={isRequired}
      className={classnames(styles.selectMenu, className)}
      {...rest}
    >
      {({ open, value }) => (
        <>
          <Listbox.Button
            id={id}
            ref={ref} // Pass the incoming ref to the button so react-hook-form can focus it during validation.
            aria-describedby={ariaDescribedBy}
            aria-invalid={isInvalid}
            className={buttonClassNames}
          >
            {value?.name || <>&nbsp;</>}

            <Icon size={iconSizes.small} className={styles.listBoxButtonIcon}>
              {open ? <ChevronUpIcon /> : <ChevronDownIcon />}
            </Icon>
          </Listbox.Button>

          <Transition
            as={Fragment}
            unmount={false}
            show={open}
            enter="transition ease-out duration-150"
            enterFrom="opacity-0 translate-y-1"
            enterTo="opacity-100 translate-y-0"
            leave="transition ease-in duration-150"
            leaveFrom="opacity-100 translate-y-0"
            leaveTo="opacity-0 translate-y-1"
          >
            <Listbox.Options
              static
              className={classnames(styles.listBoxOptions, {
                [styles.scrollable]: isScrollable,
                [styles.fitContent]: isFitContent
              })}
            >
              {['', ...options].map(renderOption)}
            </Listbox.Options>
          </Transition>
        </>
      )}
    </Listbox>
  );
});

function renderOption(option, index) {
  return (
    <Listbox.Option
      key={option?.id || index}
      value={option}
      disabled={option?.disabled}
      className={classnames(styles.listBoxOption, {
        [styles.separator]: option?.separator === true
      })}
    >
      {({ selected }) => (
        <>
          {selected && (
            <Icon size={iconSizes.small} className={styles.selectedOptionIcon}>
              <CheckmarkIcon />
            </Icon>
          )}

          {option?.separator ? <></> : <>{option?.name || <>&nbsp;</>}</>}
        </>
      )}
    </Listbox.Option>
  );
}

SelectMenu.propTypes = {
  id: PropTypes.string.isRequired,
  name: PropTypes.string,
  value: PropTypes.oneOfType([
    PropTypes.string,
    PropTypes.shape({
      id: PropTypes.string,
      name: PropTypes.string,
      value: PropTypes.string
    })
  ]),
  options: PropTypes.arrayOf(
    PropTypes.shape({
      id: PropTypes.string.isRequired,
      name: PropTypes.string.isRequired,
      value: PropTypes.string.isRequired,
      disabled: PropTypes.bool,
      separator: PropTypes.bool
    })
  ),
  isRequired: PropTypes.bool,
  isDisabled: PropTypes.bool,
  isInvalid: PropTypes.bool,
  ariaDescribedBy: PropTypes.string,
  onBlur: PropTypes.func,
  onChange: PropTypes.oneOfType([
    PropTypes.func,
    PropTypes.arrayOf(PropTypes.func)
  ]),
  colorScheme: colorSchemePropType,
  className: PropTypes.string,
  isScrollable: PropTypes.bool,
  isFitContent: PropTypes.bool
};

```

**Comment by TyreeceSimpson** [comment]
I fixed a very similar issue by using the 'by' prop on the listbox to compare the objects by a particular field instead of comparing object identity which will be different if it is not the exact same instance.
[https://headlessui.com/react/listbox#binding-objects-as-values](https://headlessui.com/react/listbox#binding-objects-as-values)
